internal/affine/colorm.go

```diff
@@ -167,6 +167,229 @@ func (c *ColorM) UnsafeElements() ([]float32, []float32) {
 	return eb, et
 }
 
+func (c *ColorM) det() float32 {
+	if c == nil {
+		return 0
+	}
+	if !c.isInited() {
+		return 1
+	}
+
+	m00 := c.body[0]
+	m01 := c.body[1]
+	m02 := c.body[2]
+	m03 := c.body[3]
+	m04 := float32(0)
+	m10 := c.body[4]
+	m11 := c.body[5]
+	m12 := c.body[6]
+	m13 := c.body[7]
+	m14 := float32(0)
+	m20 := c.body[8]
+	m21 := c.body[9]
+	m22 := c.body[10]
+	m23 := c.body[11]
+	m24 := float32(0)
+	m30 := c.body[12]
+	m31 := c.body[13]
+	m32 := c.body[14]
+	m33 := c.body[15]
+	m34 := float32(0)
+	m40 := c.translate[0]
+	m41 := c.translate[1]
+	m42 := c.translate[2]
+	m43 := c.translate[3]
+	m44 := float32(1)
+
+	a3434 := m33*m44 - m34*m43
+	a2434 := m32*m44 - m34*m42
+	a2334 := m32*m43 - m33*m42
+	a1434 := m31*m44 - m34*m41
+	a1334 := m31*m43 - m33*m41
+	a1234 := m31*m42 - m32*m41
+	a0434 := m30*m44 - m34*m40
+	a0334 := m30*m43 - m33*m40
+	a0234 := m30*m42 - m32*m40
+	a0134 := m30*m41 - m31*m40
+
+	b234234 := m22*a3434 - m23*a2434 + m24*a2334
+	b134234 := m21*a3434 - m23*a1434 + m24*a1334
+	b124234 := m21*a2434 - m22*a1434 + m24*a1234
+	b123234 := m21*a2334 - m22*a1334 + m23*a1234
+	b034234 := m20*a3434 - m23*a0434 + m24*a0334
+	b024234 := m20*a2434 - m22*a0434 + m24*a0234
+	b023234 := m20*a2334 - m22*a0334 + m23*a0234
+	b014234 := m20*a1434 - m21*a0434 + m24*a0134
+	b013234 := m20*a1334 - m21*a0334 + m23*a0134
+	b012234 := m20*a1234 - m21*a0234 + m22*a0134
+
+	det := m00*(m11*b234234-m12*b134234+
+		m13*b124234-m14*b123234) -
+		m01*(m10*b234234-m12*b034234+
+			m13*b024234-m14*b023234) +
+		m02*(m10*b134234-m11*b034234+
+			m13*b014234-m14*b013234) -
+		m03*(m10*b124234-m11*b024234+
+			m12*b014234-m14*b012234) +
+		m04*(m10*b123234-m11*b023234+
+			m12*b013234-m13*b012234)
+	if det == 0 {
+		return 0
+	}
+	return 1 / det
+}
+
+// IsInvertible returns a boolean value indicating
+// whether the matrix c is invertible or not.
+func (c *ColorM) IsInvertible() bool {
+	return c.det() != 0
+}
+
+// Invert inverts the matrix.
+// If c is not invertible, Invert panics.
+func (c *ColorM) Invert() {
```

Comment:
As ColorM is an affine matrix, you don't have to calculate inversion of a 5x5 matrix. Calculating an inversion of 4x4 matrix is enough! Assuming that M is an affine 5x5 matrix that consists of M_body (5x5 only with the body) and M_translate (5x5 only with the translate), M^(-1) can be calculated like this:

```
M^(-1) = (M_body * M_translate)^(-1)
       = M_translate^(-1) * M_body^(-1)
```

`M_translate^(-1)` can be calculated easily: just multiply -1 to M_translate (imagine that moving something by an offset X. The opposite operation is to move it by -X).

`M_body^(-1)` can be calculated in the same way of inverting 4x4 matrix, even though M_body is a 5x5 matrix :-)

BTW, I realized that GeoM's det doesn't actually calculate 3x3 det, so I renamed this :-) 5bc09acb41fa10a1b436780daeaaf034b15483f6

Reply:
I assumed that ColorM was always a 5x5 by reading the issue lol. Initially I generated a 4x4 inversion, but by inspecting the `SetElement(i, j int, element float32)`, the `c.translate` seemed to be the last row of the 5x5 matrix to invert, then I also assumed that the last column had to exist at the net and Invert funcs (`0 0 0 0 1`).

So, the best way to do this is to generate a 4x4 inversion for `c.body` and just multiply the `c.translate` by -1?

Reply:
> ... I also assumed that the last column had to exist at the det and Invert funcs (0 0 0 0 1).

Yes, this assumption is correct. Your calculation seems totally fine, but I'm suggesting a more efficient way.

> So, the best way to do this is to generate a 4x4 inversion for c.body and just multiply the c.translate by -1?

Very close. As I indicated, the multiplying order is `M_translate^(-1) * M_body^(-1)`.

Reply:
I removed all the unnecessary multiplications (by the last column 0 0 0 0 1) and now the `det` is 4x4. The inversion still uses the `c.translate`, but since the last column was removed, it's more efficient.

Reply:
Benchmarks:
```
goos: darwin
goarch: amd64
pkg: github.com/hajimehoshi/ebiten/internal/affine
BenchmarkColorMInvert-4   	 5123756	       215 ns/op	     128 B/op	       3 allocs/op
PASS
ok  	github.com/hajimehoshi/ebiten/internal/affine	1.527s
```